File: Java/MedianOfMedians.java
import java.util.Arrays;
import sorting.QuickSort;

public class MedianOfMedians {
	public static void main(String[] args) {
		int[] arr = {7,4,9,6,5,3,8,10,17};
		System.out.println("Original Array: "+Arrays.toString(arr));
		
		int median = median_of_medians(arr);

		System.out.println("Sorted Array: "+Arrays.toString(QuickSort.quick_sort(arr,0,arr.length-1)));
		System.out.println("Median: "+median);
	}

	public static int median_of_medians(int[] arr) {
		int num_of_arr = (int)Math.ceil((double)arr.length/5);
		int[][] subarr = new int[num_of_arr][];

		int last_size = ( arr.length%5 == 0 ) ? 5 : arr.length%5;
		int size_of_arr = 5;

		for ( int i=0;i<num_of_arr;i++ ) {
			if ( i == num_of_arr-1 ) size_of_arr = last_size;
			// System.out.println(""+size_of_arr);
			subarr[i] = new int[size_of_arr];
			System.arraycopy(arr,i*5,subarr[i],0,size_of_arr);
			subarr[i] = QuickSort.quick_sort(subarr[i],0,size_of_arr-1);
		}

		// System.out.println("Original Array: "+Arrays.deepToString(subarr));

		int[] median_arr = find_median(subarr);
		// System.out.println("Median Array: "+Arrays.toString(median_arr));
		
		int l_m = median_arr.length;
		
		
		if ( l_m <= 5 ) return QuickSort.quick_sort(median_arr,0,l_m-1)[l_m/2];
		else return median_of_medians(median_arr);
	}

	public static int[] find_median(int[][] subarr) {
		int l = subarr.length;
		int[] median = new int[l];
		for ( int i=0;i<l;i++ ) {
			median[i] = subarr[i][subarr[i].length/2];
		}

		return median;
	} 
}

File: Python/median_of_medians.py
import time
start_time = time.time()
# import sys
# sys.path.insert(0, 'Sorting')
# from quick_sort import quick_sort

def median_of_medians(arr):

	# runs in --- seconds --- 0.00699996948242
	# better one (copied from net)

	# divide A into sublists of len 5
    sublists = [arr[j:j+5] for j in range(0, len(arr), 5)]
    medians = [sorted(sublist)[len(sublist)/2] for sublist in sublists]
    if len(medians) <= 5:
        return sorted(medians)[len(medians)/2]
    else:
        return median_of_medians(medians)


#   # runs in ---seconds --- 0.00800013542175
#  	# mine one
	
# 	subarr = []
# 	l = len(arr)
# 	i=0

# 	# divide A into sublists of len 5
# 	while l > 0:
# 		if l < 5:
# 			end_index = i+l
# 		else:
# 			end_index = i+5
# 		temp = arr[i:end_index]
# 		subarr.append(quick_sort(temp,0,len(temp)-1))
# 		i = i+5
# 		l = l - 5
# 	# print subarr
	
# 	median_arr = find_median(subarr)
# 	# print median_arr

# 	l_m = len(median_arr)
# 	if l_m <= 5:
# 		return quick_sort(median_arr,0,l_m-1)[l_m/2]
# 	else:
# 		return median_of_medians(median_arr)
	

# def find_median(subarr):
# 	median = []
# 	for i in range(len(subarr)):
# 		median.append(subarr[i][len(subarr[i])/2])

# 	return median

if __name__ == '__main__':
	arr = [7,4,9,6,5,3,19,37,1,7,4,6,50,33,23,45,67,89,56,23,45,6,8,9,10,13,12,17,16,18,19,20,33,37]
	print 'Original: ',arr
	print 'Sorted: ',sorted(arr)
	# print 'Sorted: ',quick_sort(arr,0,len(arr)-1)
	median = median_of_medians(arr)
	print 'Median: ',median
	print "---seconds ---", (time.time()-start_time)

